dotnet/src/Connectors/Connectors.AI.OpenAI/ChatCompletionWithData/AzureChatCompletionWithData.cs

```diff
@@ -0,0 +1,345 @@
+
+    private const string ServerEventPayloadPrefix = "data:";
+    private const string DefaultApiVersion = "2023-06-01-preview";
+    private const string EndpointUriFormat = "{0}/openai/deployments/{1}/extensions/chat/completions?api-version={2}";
+
+    private readonly AzureChatCompletionWithDataConfig _config;
+
+    private readonly HttpClient _httpClient;
+    private readonly ILogger _logger;
+
+    private void ValidateConfig(AzureChatCompletionWithDataConfig config)
+    {
+        Verify.NotNull(config);
+
+        Verify.NotNullOrWhiteSpace(config.CompletionModelId);
+        Verify.NotNullOrWhiteSpace(config.CompletionEndpoint);
+        Verify.NotNullOrWhiteSpace(config.CompletionApiKey);
+        Verify.NotNullOrWhiteSpace(config.DataSourceEndpoint);
+        Verify.NotNullOrWhiteSpace(config.DataSourceApiKey);
+        Verify.NotNullOrWhiteSpace(config.DataSourceIndex);
+    }
+
+    private static void ValidateMaxTokens(int? maxTokens)
+    {
+        if (maxTokens.HasValue && maxTokens < 1)
+        {
+            throw new AIException(
+                AIException.ErrorCodes.InvalidRequest,
+                $"MaxTokens {maxTokens} is not valid, the value must be greater than zero");
+        }
+    }
+
+    private async Task<IReadOnlyList<IChatResult>> ExecuteCompletionRequestAsync(
+        ChatHistory chat,
+        ChatRequestSettings requestSettings,
+        CancellationToken cancellationToken = default)
+    {
+        var request = this.GetRequest(chat, requestSettings, isStreamEnabled: false);
+        using var response = await this.GetResponse(request, cancellationToken).ConfigureAwait(false);
+
+        this.EnsureSuccessStatusCode(response);
+
+        var body = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+
+        var chatWithDataResponse = this.DeserializeResponse<ChatWithDataResponse>(body);
+
+        return chatWithDataResponse.Choices.Select(choice => new ChatWithDataResult(chatWithDataResponse, choice)).ToList();
+    }
+
+    private async IAsyncEnumerable<IChatStreamingResult> ExecuteCompletionStreamingRequestAsync(
+        ChatHistory chat,
+        ChatRequestSettings requestSettings,
+        [EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        var request = this.GetRequest(chat, requestSettings, isStreamEnabled: true);
+        using var response = await this.GetResponse(request, cancellationToken).ConfigureAwait(false);
+
+        this.EnsureSuccessStatusCode(response);
+
+        await foreach (var result in this.GetStreamingResultsAsync(response))
+        {
+            yield return result;
+        }
+    }
+
+    private async Task<HttpResponseMessage> GetResponse(
+        ChatWithDataRequest request,
+        CancellationToken cancellationToken = default)
+    {
+        using var httpRequestMessage = HttpRequest.CreatePostRequest(this.GetRequestUri(), request);
```

Comment:
nit: The code for creating an HTTP request message can be moved to the GetRequest method, and this method can be renamed to SendRequestAsync to better align method names with what they actually do.

Reply:
Good suggestion, thanks! Code updated.